src/permissions.cxx

```diff
@@ -0,0 +1,174 @@
+#include "../include/permissions.hxx"
+#include <fmt/core.h>
+#include <fmt/ranges.h>
+
+// TODO: It might be useful to add an optional condition which must be
+// met for conditions to apply. So a variable is only read or written
+// if another variable is already set, for example. Could potentially
+// have conditions based on values as well. Could put in boolean
+// logic. This would allow finer-grained access control in Components,
+// but is it worth the hassle?
+//
+// struct Condition {
+//   virtual bool eval() const = 0;
+// }
+//
+// struct IsSetCondition : Condition {
+//   std::string varname
+// }
+//
+// struct TrueCondition;
+// struct AndCondition;
+// struct OrCondition;
+//
+// There should be ways to do this with templates to allow greater inlining
+//
+
+/// Return a set of access rights where the lower permissions have
+/// been updated so that they reflect higher permissions (e.g., read
+/// permission will be set in all cases where write permission was
+/// set).
+Permissions::AccessRights applyLowerPermissions(const Permissions::AccessRights& rights) {
+  Permissions::AccessRights result(rights);
+  for (int i = static_cast<int>(PermissionTypes::ReadIfSet);
+       i < static_cast<int>(PermissionTypes::END); i++) {
+    // Higher permissions imply lower permissions
+    for (int j = static_cast<int>(PermissionTypes::ReadIfSet); j < i; j++) {
+      result[j] = result[j] | rights[i];
+    }
+  }
+  return result;
+}
+
+const std::map<Regions, std::string> Permissions::fundamental_regions = {
+    {Regions::Interior, "Interior"}, {Regions::Boundaries, "Boundaries"}};
+
+Permissions::Permissions(std::initializer_list<VarRights> data) : variable_permissions() {
+  for (const auto& [varname, access] : data) {
+    setAccess(varname, access);
+  }
+}
+
+void Permissions::setAccess(const std::string& variable, const AccessRights& rights) {
+
+  variable_permissions[variable] = applyLowerPermissions(rights);
+}
+
+std::string replaceAll(const std::string& str, const std::string& from,
+                       const std::string& to) {
+  std::string result = str;
+  if (from.empty())
+    return result;
+  size_t start_pos = 0;
+  while ((start_pos = result.find(from, start_pos)) != std::string::npos) {
+    result.replace(start_pos, from.length(), to);
+    start_pos +=
+        to.length(); // In case 'to' contains 'from', like replacing 'x' with 'yx'
+  }
+  return result;
+}
+
+void Permissions::printAllPermissions() const {
+  for (const auto& [varname, perms] : variable_permissions) {
+    fmt::print("{}", varname);
+    for (int i = 0; i < static_cast<int>(PermissionTypes::END); i++) {
+      fmt::print(" {}", static_cast<int>(perms[i]));
+    }
+    fmt::print("\n");
```

Comment:
Is this just:
```suggestion
    fmt::print("{} {}\n", varname, fmt::join(" ", perms));
```

Alternatively, replace this function with an fmt formatter, then it's possible to convert to a string and use in more contexts